enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj

```diff
@@ -0,0 +1,463 @@
+                        :indexer_last_seen Instant/EPOCH}
+        step-ex        (volatile! nil)                      ; holds an exception to be thrown during the step function
+        indexing-state (semantic.indexer/init-indexing-state metadata-row)
+        call-counter   (atom {:idle 0 :step 0})
+        wait-for-calls (fn []
+                         (let [ocalls   @call-counter
+                               max-wait (+ (System/currentTimeMillis) 100)]
+                           (loop []
+                             (cond
+                               (not= ocalls @call-counter) true
+                               (< max-wait (System/currentTimeMillis)) false
+                               :else (recur)))))
+        step-called    (promise)
+        idle-called    (promise)]
+
+    (with-redefs [semantic.indexer/on-indexing-idle (fn [_]
+                                                      (deliver idle-called true)
+                                                      (swap! call-counter update :idle inc)
+                                                      (when (.isInterrupted (Thread/currentThread))
+                                                        (throw (InterruptedException.))))
+                  semantic.indexer/indexing-step    (fn [& _]
+                                                      (deliver step-called true)
+                                                      (swap! call-counter update :step inc)
+                                                      (when-some [ex @step-ex] (throw ex))
+                                                      (when (.isInterrupted (Thread/currentThread))
+                                                        (throw (InterruptedException.))))]
+
+      (with-open [loop-thread
+                  (open-loop-thread! pgvector
+                                     index-metadata
+                                     index
+                                     indexing-state)]
+        (let [{:keys [caught-ex thread]} @loop-thread]
+          (is (not= :timeout (deref idle-called 100 :timeout)))
+          (is (not= :timeout (deref step-called 100 :timeout)))
+
+          (testing "start time set"
+            (is (inst? (:start-time @indexing-state))))
+
+          (testing "called repeatedly"
+            (is (wait-for-calls)))
+
+          (testing "called 1 to 1"
+            (let [{:keys [idle step]} @call-counter
+                  lower (dec idle)
+                  upper (inc idle)]
+              (is (<= lower step upper))))
+
+          (testing "exceptions are bubbled out of loop, will not loop forever"
+            (let [ex (Exception. (str "Error: " (u/generate-nano-id)))]
+              (vreset! step-ex ex)
+              (wait-for-calls)
+              (is (= (ex-message ex) (ex-message @caught-ex)))
+              (vreset! step-ex nil)
+              (wait-for-calls)
+              (testing "sanity: check loop no longer throwing"
```

Comment:
redundant, delete (this survived a previous pass when I simulated the quartz job here a bit more). 

We should `(is)` check the returns of `(wait-for-calls)` otherwise it just times out and the error is not reported. Go through and fix these cases where the return is ignored.

Reply:
107c573e4159